### DISCUSSION TITLE
Using a checkbox with a string value will return an unintended value.

### DISCUSSION THREAD

**Discussion by shota-murakami** [react-hook-form/react-hook-form#6477] [Q&A] [answered]
When I submit, I expect an array string, but booleans are mixed.

What if I want to get only the checked strings?

Here is the sample code.

https://codesandbox.io/s/hopeful-rgb-8tp9k?file=/src/App.tsx

![image](https://user-images.githubusercontent.com/18521362/132284144-4b904258-883b-4aaf-adf8-cbf4223b7769.png)
In this case, arrayValue should be returned as ["value1"]

But I get ["value1", false]

**Comment by bluebill1049** [comment, marked as answer]
This is expected behavior: 

- if checkbox is without value, we will return `boolean` 
- if checkbox is provided with `value`, we will return the check `value`

so what if you want to customize that return, you should always build a controlled component and host your own local state.

https://codesandbox.io/s/sleepy-leftpad-f073s?file=/src/App.tsx

```tsx
import * as React from "react";
import { useForm, Controller, ControllerRenderProps } from "react-hook-form";
import "./styles.css";

const arrayValues = [
  {
    label: "label1",
    value: "value1"
  },
  {
    label: "label2",
    value: "value2"
  }
];

interface Form {
  arrayValue: string[];
}

const Checkboxs = ({
  onChange,
  value
}: Omit<ControllerRenderProps<Form>, "ref" | "onBlru">) => {
  const [values, setValues] = React.useState(value);

  return (
    <>
      {arrayValues.map((item, index) => {
        return (
          <label key={index}>
            <input
              type="checkbox"
              value={item.value}
              checked={!!values[index]}
              onChange={(e) => {
                const clone = [...values];
                clone[index] = e.target.value;
                onChange(clone);
                setValues(clone);
              }}
            />
            {item.label}
          </label>
        );
      })}
    </>
  );
};

export default function App() {
  const { control, handleSubmit } = useForm<Form>({
    defaultValues: {
      arrayValue: ["", ""]
    }
  });

  const submit = (data: any) => {
    console.log(data);
  };

  return (
    <div className="App">
      <form onSubmit={handleSubmit(submit)}>
        <Controller
          name="arrayValue"
          control={control}
          render={({ field: { ref, ...rest } }) => <Checkboxs {...rest} />}
        />
        <button type="submit">submit</button>
      </form>
    </div>
  );
}
```

**Reply by shota-murakami** [reply]
thank you for your answer!!
If you want to get only the selected value. For example, in this code, if you want to set array Value: ["value1"], need to perform filter processing at such as submit timing?

**Reply by bluebill1049** [reply]
you should move that logic into the submit callback. 👍

**Reply by shota-murakami** [reply]
Thank you for the quick response! It was very easy to understand
